radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,746 @@
+        &mut self,
+        title: String,
+        description: String,
+        signer: &G,
+    ) -> Result<OpId, Error> {
+        self.transaction("Edit", signer, |tx| tx.edit(title, description))
+    }
+
+    /// Publish a proposal.
+    pub fn publish<G: Signer>(&mut self, signer: &G) -> Result<OpId, Error> {
+        self.transaction("Publish", signer, |tx| tx.publish())
+    }
+
+    /// Close a proposal.
+    pub fn close<G: Signer>(&mut self, signer: &G) -> Result<OpId, Error> {
+        self.transaction("Close", signer, |tx| tx.close())
+    }
+
+    /// Comment on a proposal revision.
+    pub fn comment<G: Signer, S: ToString>(
+        &mut self,
+        revision: RevisionId,
+        body: S,
+        reply_to: thread::CommentId,
+        signer: &G,
+    ) -> Result<thread::CommentId, Error> {
+        self.transaction("Comment", signer, |tx| tx.comment(revision, body, reply_to))
+    }
+
+    /// Update a proposal with new metadata.
+    pub fn update<G: Signer>(
+        &mut self,
+        title: impl ToString,
+        description: impl ToString,
+        current: impl Into<Oid>,
+        proposed: Doc<Verified>,
+        signer: &G,
+    ) -> Result<(OpId, OpId), Error> {
+        self.transaction("Add revision", signer, |tx| {
+            let revision = tx.revision(current.into(), proposed);
+            let meta = tx.edit(title, description);
+
+            (revision, meta)
+        })
+    }
+
+    /// Rebase a proposal with the new `current`.
+    pub fn rebase<G: Signer>(
+        &mut self,
+        current: impl Into<Oid>,
+        proposed: Doc<Verified>,
+        signer: &G,
+    ) -> Result<OpId, Error> {
+        self.transaction("Rebase revision", signer, |tx| {
+            tx.revision(current.into(), proposed)
+        })
+    }
+}
+
+impl<'a, 'g> Deref for ProposalMut<'a, 'g> {
+    type Target = Proposal;
+
+    fn deref(&self) -> &Self::Target {
+        &self.proposal
+    }
+}
+
+pub struct Proposals<'a> {
+    raw: store::Store<'a, Proposal>,
+}
+
+impl<'a> Deref for Proposals<'a> {
+    type Target = store::Store<'a, Proposal>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.raw
+    }
+}
+
+impl<'a> Proposals<'a> {
+    /// Open an patches store.
+    pub fn open(
+        whoami: PublicKey,
+        repository: &'a storage::Repository,
+    ) -> Result<Self, store::Error> {
+        let raw = store::Store::open(whoami, repository)?;
+
+        Ok(Self { raw })
+    }
+
+    /// Create a patch.
+    pub fn create<'g, G: Signer>(
+        &'g mut self,
+        title: impl ToString,
+        description: impl ToString,
+        current: impl Into<Oid>,
+        proposed: Doc<Verified>,
+        signer: &G,
+    ) -> Result<ProposalMut<'a, 'g>, Error> {
+        let (id, patch, clock) =
+            Transaction::initial("Create proposal", &mut self.raw, signer, |tx| {
+                tx.revision(current.into(), proposed);
+                tx.edit(title, description);
+            })?;
```

Comment:
This is the only case where I'd keep both revision & edit in the same tx/function.